cpp/include/cugraph/visitors/bfs_visitor.hpp

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Andrei Schaffer, aschaffer@nvidia.com
+//
+
+#pragma once
+#include "erased_pack.hpp"
+#include "graph_envelope.hpp"
+#include "ret_terased.hpp"
+
+namespace cugraph {
+namespace experimental {
+
+// macro option: MAKE_VISITOR(bfs)
+
+// primary empty template:
+//
+template <typename vertex_t,
+          typename edge_t,
+          typename weight_t,
+          bool st,
+          bool mg,
+          typename Enable = void>
+struct bfs_visitor;
+
+// dummy out non-candidate instantiation paths:
+//
+template <typename vertex_t, typename edge_t, typename weight_t, bool st, bool mg>
+struct bfs_visitor<vertex_t,
+                   edge_t,
+                   weight_t,
+                   st,
+                   mg,
+                   std::enable_if_t<(!is_candidate<vertex_t, edge_t, weight_t>::value)>>
+  : visitor_t {
+  void visit_graph(graph_envelope_t::base_graph_t const&) override
+  {
+    // purposely empty
+  }
+  return_t const& get_result(void) const override
+  {
+    static return_t r{};
+    return r;
+  }
+};
+
+template <typename vertex_t, typename edge_t, typename weight_t, bool st, bool mg>
+struct bfs_visitor<vertex_t,
+                   edge_t,
+                   weight_t,
+                   st,
+                   mg,
+                   std::enable_if_t<is_candidate<vertex_t, edge_t, weight_t>::value>> : visitor_t {
+  bfs_visitor(erased_pack_t& ep) : ep_(ep) {}
+
+  void visit_graph(graph_envelope_t::base_graph_t const&) override;
+
+  return_t const& get_result(void) const override { return result_; }
+
+ private:
+  erased_pack_t& ep_;
+  return_t result_;
+};
+
+// wrapper:
+//
+return_t bfs_wrapper(graph_envelope_t const& g, erased_pack_t& ep);
```

Comment:
Question to consider: Does hiding all parameters in the erased_pack_t make the code more or less maintainable?

Would it make sense to have the non-type-erased parameters explicitly included in the call and put the type erased parameters in the erased_pack_t?

Reply:
Second question... should we consider defining the wrapper in a different file?  In the C++ API we define all algorithms in a single header file (`algorithms.hpp`).  Would it make sense to do the same so that the caller only needs to include a single file to access all algorithms?

Reply:
Third question... does the suffix `_wrapper` actually add value?  If we move (as I suggested in a different comment) these into a separate namespace (e.g. cugraph::api), couldn't this just be `cugraph::api::bfs`?

Reply:
Question 1: See answer here: https://github.com/rapidsai/cugraph/pull/1711#discussion_r670572920;

Question 2: Yes. Once I add at least one more wrapper I will move all into their own header.

Question 3: It probably does not (add any value). I think moving this into its own namespace and calling it by its proper name (`bfs()`) make sense. I'm not sure what to call this separate namespace, though. Maybe `namespace wrappers`?

Reply:
Question 3 follow-up: yes, `cugraph::api` should work, but as said I'm planning to also add a `cugraph/api` directory. I'm not sure if that would confuse things.

Reply:
Done: Isolated API in its own header; added `cugraph::api` namespace;